portal-bridge/src/types/state/execution.rs

```diff
@@ -0,0 +1,168 @@
+use alloy_primitives::{keccak256, Address, Bytes, B256, U256};
+use alloy_rlp::Decodable;
+use anyhow::{ensure, Error};
+use eth_trie::{EthTrie, MemoryDB, Trie};
+use ethportal_api::types::state_trie::account_state::AccountState;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeSet, HashMap},
+    fs::File,
+    io::BufReader,
+    sync::Arc,
+};
+
+use crate::types::{era1::BlockTuple, state::blockreward::get_block_reward};
+
+#[derive(Debug, Serialize, Deserialize)]
+struct AllocBalance {
+    balance: U256,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct GenesisConfig {
+    alloc: HashMap<Address, AllocBalance>,
+    #[serde(rename = "stateRoot")]
+    state_root: B256,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct AccountProof {
+    pub address: Address,
+    pub state: AccountState,
+    pub proof: Vec<Bytes>,
+}
+
+#[warn(dead_code)]
+pub struct State {
+    pub accounts: BTreeSet<Address>,
+    trie: Box<EthTrie<MemoryDB>>,
+    next_block_number: u64,
+}
+
+impl State {
+    pub fn new() -> anyhow::Result<Self> {
+        let mut trie = EthTrie::new(Arc::new(MemoryDB::new(false)));
+        let mut accounts = BTreeSet::new();
+
+        let genesis_file = File::open("src/types/state/mainnet.json")?;
+        let genesis: GenesisConfig = serde_json::from_reader(BufReader::new(genesis_file))?;
+
+        for (address, alloc_balance) in genesis.alloc {
+            accounts.insert(address);
+
+            let mut account = AccountState::default();
+            account.balance += alloc_balance.balance;
+            trie.insert(keccak256(address).as_ref(), &alloy_rlp::encode(account))?;
+        }
+
+        let root = B256::from_slice(trie.root_hash()?.as_slice());
+        ensure!(
+            root == genesis.state_root,
+            "Root doesn't match state root from genesis file"
+        );
+        Ok(State {
+            accounts,
+            trie: Box::from(trie),
+            next_block_number: 1,
+        })
+    }
+
+    pub fn process_block(&mut self, block_tuple: &BlockTuple) -> anyhow::Result<BTreeSet<Address>> {
+        ensure!(
+            self.next_block_number == block_tuple.header.header.number,
+            "Expected block {}, received {}",
+            self.next_block_number,
+            block_tuple.header.header.number
+        );
+
+        let mut updated_accounts = BTreeSet::new();
+        // execute transactions
+        let transactions = match block_tuple.body.body.transactions() {
+            Ok(transactions) => transactions,
+            Err(e) => return Err(Error::msg(format!("Error getting transactions: {e:?}"))),
+        };
+        for _tranasction in transactions {}
```

Comment:
?

Reply:
This is placeholder for the follow up PR